tests/integration/test_scan.py

```diff
@@ -0,0 +1,168 @@
+# limitations under the License.
+#
+
+
+import numpy as np
+import pytest
+
+import cunumeric as num
+
+np.random.seed(12345)
+
+
+def _gen_array(n0, shape, dt, axis, outtype):
+    # range 1-10, avoiding zeros to ensure correct testing for int prod case
+    if dt == np.complex64:
+        A = (
+            (99 * np.random.random(shape) + 1)
+            + (99 * np.random.random(shape) + 1) * 1j
+        ).astype(np.complex64)
+    elif dt == np.complex128:
+        A = (
+            (99 * np.random.random(shape) + 1)
+            + (99 * np.random.random(shape) + 1) * 1j
+        ).astype(np.complex128)
+    else:
+        A = (99 * np.random.random(shape) + 1).astype(dt)
+    if n0 == "first_half":
+        # second element along all axes is a NAN
+        A[(1,) * len(shape)] = np.nan
+    elif n0 == "second_half":
+        # second from last element along all axes is a NAN
+        A[
+            tuple(map(lambda i, j: i - j, A.shape, (2,) * len(A.shape)))
+        ] = np.nan
+    if outtype is None:
+        B = None
+        C = None
+    else:
+        if axis is None:
+            B = np.zeros(shape=A.size, dtype=outtype)
+            C = np.zeros(shape=A.size, dtype=outtype)
+        else:
+            B = np.zeros(shape=shape, dtype=outtype)
+            C = np.zeros(shape=shape, dtype=outtype)
+    return A, B, C
+
+
+def _run_tests(op, n0, shape, dt, axis, out0, outtype):
+    print(
+        f"Running test: {op}, shape: {shape}, nan location: {n0}"
+        f", axis: {axis}, in type: {dt}, out type: {outtype}"
+        f", output array not provided: {out0}"
+    )
+    if out0 is True:
+        A, B, C = _gen_array(n0, shape, dt, axis, None)
+        B = getattr(num, op)(A, out=None, axis=axis, dtype=outtype)
+        C = getattr(np, op)(A, out=None, axis=axis, dtype=outtype)
+    else:
+        A, B, C = _gen_array(n0, shape, dt, axis, outtype)
+        B = getattr(num, op)(A, out=B, axis=axis, dtype=outtype)
+        C = getattr(np, op)(A, out=C, axis=axis, dtype=outtype)
+
+    print("Checking result...")
+    if np.allclose(B, C, equal_nan=True):
+        print("PASS!")
+    else:
+        print("FAIL!")
+        print(f"INPUT    : {A}")
+        print(f"CUNUMERIC: {B}")
+        print(f"NUMPY    : {C}")
+        assert False
+
+
+ops = [
+    "cumsum",
+    "cumprod",
+    "nancumsum",
+    "nancumprod",
+]
+# keeping array sizes small to avoid accumulation variance
+# between cunumeric and numpy
+shapes = [
+    [10000],
+    [10, 1000],
+    [10, 10, 100],
+]
+axes = [
+    None,
+    0,
+]
+out0s = [
+    True,
+    False,
+]
+
+
+@pytest.mark.parametrize("op", ops)
+@pytest.mark.parametrize("shape", shapes)
+@pytest.mark.parametrize("axis", axes)
+@pytest.mark.parametrize("out0", out0s)
+def test_scan(op, shape, axis, out0):
+    n0s = [
+        None,
+        "first_half",
+        "second_half",
+    ]
+    int_types = [
+        np.int32,
+        np.int64,
+    ]
+    float_types = [
```

Comment:
looks like your code doesn't support `float16`. is that right? I think you should add support for it.

Reply:
9ffdb020deed2b5e1a217c5f248b8f7cca4d874a